docs/types/dynamic-dispatch.md

```diff
@@ -162,3 +162,137 @@ main = Test.simplify
 When invoking a method on _module object_ its _module static methods_ take
 precedence over _instance methods_ defined on `Any`. Thus a module serves
 primarily as a _container for module (static) methods_.
+
+## Method Resolution
+
+This section describes the _method resolution_ process, which resolves a
+concrete method definition for a concrete call site. For a method call
+expression `Receiver.symbol`, this section focuses only on a single dispatch
+based on `Receiver` argument. For multiple dispatch, see the
+[Multiple Dispatch](#multiple-dispatch) section.
+
+Method resolution algorithm for `Receiver.symbol`:
+
+1. **Determine the kind of `Receiver`:**
+
+   - 1.1. If `Receiver` is a type (type object), go to step 2.
+   - 1.2. If `Receiver` is a value (instance / atom), go to step 3.
+   - 1.3. If `Receiver` is a module, go to step 4.
+   - 1.4. If `Receiver` is a polyglot object, go to step 5.
+   - 1.5. If there is no `Receiver`, go to step 6.
+
+2. **`Receiver` is a type. This call site is a _static method call_:**
+
+   - 2.1. We are looking for either a _static method_ or an _instance method_
+     defined on `Receiver` type.
+   - 2.2. If `symbol` is defined on `Receiver`, with or without `self` argument,
+     select it and stop.
+   - 2.3. If `symbol` is defined on `Any`, with or without `self` argument,
+     select it and stop.
+   - 2.4. Otherwise, raise `No_Such_Method` panic and stop.
+
+3. **`Receiver` is a value (instance / atom):**
+
+   - 3.1. We are looking for an _instance method_.
+   - 3.2. If `symbol` is defined on `Receiver` with `self` argument, select it
+     and stop.
+   - 3.3. If `symbol` is defined on `Any` with `self` argument, select it and
+     stop.
+   - 3.4. Otherwise, raise `No_Such_Method` panic and stop.
+
+4. **`Receiver` is a module. We are looking for a _module method_:**
+
+   - 4.1. Module methods are not allowed to be defined with `self` argument.
+   - 4.2. If `symbol` is defined in `Receiver` module, select it and stop.
+   - 4.3. Otherwise, raise `No_Such_Method` panic and stop.
+
+5. **`Receiver` is a polyglot object. A polyglot object can be:**
+
+   - 5.1. A Java class, imported by `polyglot java import ...` statement.
+   - 5.2. Java object instance, created by `Java_Class.new ...` expression.
+   - 5.3. Javascript, Python, or any other allowed foreign language object
+     returned by a foreign method call.
+   - 5.4. Searching for `symbol` on such a `Receiver` is subject to the rules of
+     the [polyglot Interoperability](../polyglot/README.md) chapter.
+
+6. **There is no `Receiver`:**
+   - 6.1. We are looking either for a _module method_ or for a variable in the
```

Comment:
- Not really a "method resolution"
- this is only influenced by "lexical scope" with all the defined symbols being available
- this should even yield an error during "compile time" - e.g. before executing any code